app/controllers/instructeurs/procedures_controller.rb

```diff
@@ -324,8 +325,48 @@ def apercu
       DossierPreloader.load_one(@dossier)
     end
 
+    def history
+      @procedure = procedure
+      all_published_revisions = @procedure.revisions.where.not(published_at: nil).order(published_at: :asc)
+
+      @comparison_pairs = []
+
+      if all_published_revisions.size >= 2
+        1.upto(all_published_revisions.size - 1) do |i|
+          previous = all_published_revisions[i - 1]
+          current = all_published_revisions[i]
+
+          @comparison_pairs << [current, previous]
+        end
+
+        @comparison_pairs.reverse!
+      end
+
+      @instructeur_procedure = InstructeursProcedure.find_by(
+        instructeur: current_instructeur,
+        procedure: @procedure
+      )
+      @last_revision_seen_id = @instructeur_procedure&.last_revision_seen_id
+    end
+
     private
 
+    def mark_latest_revision_as_seen
+      if @procedure.revisions.where.not(published_at: nil).any?
+        latest_revision = @procedure.revisions.where.not(published_at: nil).last
+        if latest_revision.present?
+          instructeur_procedure = InstructeursProcedure.find_by(
+            instructeur: current_instructeur,
+            procedure: @procedure
+          )
+
+          if instructeur_procedure.last_revision_seen_id.nil? || instructeur_procedure.last_revision_seen_id != latest_revision.id
+            instructeur_procedure.update(last_revision_seen_id: latest_revision.id)
+          end
+        end
+      end
```

Comment:
une guard clause ne suffirait pas ? (check le .any? && last pas utile). Passer par un find_or_create_by() pour pas planter, et juste chainer l'update qui se declenchera si l'id change non ?

```suggestion
      return if @procedure.published_revision.blank?
      
      InstructeursProcedure
        .find_by_or_create_by(instructeur: current_instructeur, procedure: @procedure)
        .update(last_revision_seen_id: @procedure.published_revision.id)
      end
```

Reply:
enfait j'ai encore plus simple : le `mark_latest_revision_as_seen` arrivant uniquement sur le history on peut utiliser le @instructeur_procedure qu'on a setté auparavant, donc : 
```ruby
    def mark_latest_revision_as_seen
      return if @procedure.published_revision_id.blank?
      @instructeur_procedure.update(last_revision_seen_id: @procedure.published_revision_id)
    end
```